cli/listener/src/main.rs

```diff
@@ -392,7 +403,154 @@ impl Metrics {
+        .with_status_code(400)
+        .with_header(content_type)
+        .boxed()
+}
+
+/// Get an interval price, consume it and returns a `ResponseBox` with the
+/// provided interval price and computed annual percentage rate.
+fn get_success_response(interval_prices: IntervalPrices) -> ResponseBox {
+    let response_interval = ResponseInterval {
+        annual_percentage_rate: interval_prices.annual_percentage_rate(),
+        interval_prices,
+    };
+    let content_type = Header::from_bytes(&b"Content-Type"[..], &b"application/json"[..])
+        .expect("Static header value, does not fail at runtime.");
+    Response::from_data(
+        serde_json::to_vec(&response_interval).expect("Serialization shouldn't fail"),
+    )
+    .with_header(content_type)
+    .boxed()
+}
+
+/// Gets a response that will be sent to a requesting client for a price
+/// interval. Returns None if no method is passed.
+fn get_interval_price_request(
+    db_connection: &Connection,
+    request_url: &str,
+) -> Option<ResponseBox> {
+    let parsed_url = match Url::parse(request_url) {
+        Ok(parsed_url) => parsed_url,
+        Err(err) => {
+            return Some(get_error_response(ResponseError {
+                error: err.to_string(),
+            }))
+        }
+    };
+    let method_name = parsed_url.path_segments()?.last()?;
+    if method_name != "interval_price" {
+        return Some(get_error_response(ResponseError {
+                        error: format!("Method not supported: {}, use \"/interval_price?from=<from_date_iso8601>&to=<to_date_iso8601>\"", method_name)
+                    }));
+    }
+    let parsed_request_url =
+        form_urlencoded::parse(parsed_url.query()?.as_bytes()).collect::<Vec<_>>();
+
+    let dates = match get_date_params(parsed_request_url) {
+        Ok(res) => Some(res),
+        Err(err) => return Some(get_error_response(err)),
+    }?;
+
+    match (dates.start, dates.end) {
+        (Ok(from), Ok(to)) => {
+            let interval_prices = get_interval_price_for_period(
+                db_connection
+                    .unchecked_transaction()
+                    .expect("Failed to create sqlite transaction."),
+                from,
+                to,
+                SOLIDO_ID.to_owned(),
+            );
+            let interval_prices = interval_prices;
+            match interval_prices {
+                // Error while getting the interval prices.
+                Err(err) => Some(get_error_response(ResponseError {
+                    error: err.to_string(),
+                })),
+                Ok(interval_prices_opt) => {
+                    if let Some(interval_prices) = interval_prices_opt {
+                        // Got interval prices.
+                        Some(get_success_response(interval_prices))
+                    } else {
+                        // No interval price could be calculated, probably because of few data points.
```

Comment:
If we have a single data point, shall we return a success response with `APY` set to `0`, and the `begin_*` and `end_*` values set to the same value? This way the service doesn’t return errors for an entire epoch, and then we can use this endpoint for health checks.

Reply:
Yeah, can be. Since we're going to pre-populate the db with the data for when Solido was first deployed I think it wouldn't make a practical difference, but I changed it. Now it fails only if no data point is present.